### PR TITLE
feat(form-v2): preview form as admin user

### LINKED ISSUE #3363: [React/Builder] Preview form as user
My gut feeling is this is not as important, can be deprioritized for now

### PR DESCRIPTION
**Pull request by wanlingt**
## Problem
This PR adds the ability for form admins to preview their created forms

Closes #3363 

## Solution
- Add route for `PreviewFormPage`
- Add `PreviewFormPage`. `PreviewFormProvider` uses `PublicFormProvider` as reference, and uses `PublicFormContext`. Changes made to `handleSubmitForm` and `usePreviewForm`
- Add `PreviewFormHeader` as a separate component
- Modify `FormEndPageContainer` to accept `isPreview` as parameter

**Breaking Changes** 

- [x] No - this PR is backwards compatible  

## Before & After Screenshots

**AFTER**:
Desktop:
Screenshot 2022-06-24 at 11 43 15 AM

Mobile:
Screenshot 2022-06-24 at 11 43 30 AM

## Tests
- [ ] Able to preview own form as admin
- [ ] Unable to preview form if not logged in as form admin
- [ ] Preview form submission should not be reflected in Results (Responses)
- [ ] Preview form feedback submission should not be reflected in Results (Feedback)
- Added stories for `PreviewFormHeader` and `PreviewFormPage`

### CODE REVIEW

File: frontend/src/app/AppRouter.tsx [hunk 1]
```diff
@@ -90,6 +92,10 @@ export const AppRouter = (): JSX.Element => {
             />
           </Route>
         </Route>
+        <Route
+          path={`/:formId/${PREVIEW_ROUTE}`}
+          element={<PrivateElement isPreview element={<PreviewFormPage />} />}
```
Comment (karrui) [inline frontend/src/app/AppRouter.tsx]: feels a little iffy to add such a specific `isPreview` prop to such our route elements haha, should we just add the GovtMasthead directly inside `PreviewFormPage`?

File: frontend/src/app/AppRouter.tsx [hunk 2]
```diff
@@ -90,6 +92,10 @@ export const AppRouter = (): JSX.Element => {
             />
           </Route>
         </Route>
+        <Route
+          path={`/:formId/${PREVIEW_ROUTE}`}
```
Comment (karrui) [inline frontend/src/app/AppRouter.tsx]: suggest prepending `ADMINFORM_ROUTE` to the route so it is obvious that this is an admin-only route. (can't just nest inside `ADMINFORM_ROUTE` due to `AdminFormLayout` haha

File: frontend/src/app/PrivateElement.tsx [hunk 3]
```diff
@@ -1,27 +1,44 @@
...
-    element
+    isPreview ? (
+      <Flex flexDir="column" height="100vh" pos="relative">
+        <GovtMasthead />
+        <PreviewFormHeader />
+        {element}
+      </Flex>
+    ) : (
+      element
+    )
```
Comment (karrui) [inline frontend/src/app/PrivateElement.tsx]: should remove haha, inline inside `PreviewFormPage` please!

File: frontend/src/constants/routes.ts [hunk 4]
```diff
@@ -15,6 +15,7 @@ export const ADMINFORM_ROUTE = '/admin/form'
 export const ADMINFORM_BUILD_SUBROUTE = ''
 export const ADMINFORM_SETTINGS_SUBROUTE = 'settings'
 export const ADMINFORM_RESULTS_SUBROUTE = 'results'
+export const PREVIEW_ROUTE = 'preview'
```
Comment (karrui) [inline frontend/src/constants/routes.ts]:
```suggestion
export const ADMINFORM_PREVIEW_SUBROUTE = 'preview'
```

File: frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbarContainer.tsx [hunk 5]
```diff
@@ -53,7 +55,8 @@ const useAdminFormNavbar = () => {
 
   const handlePreviewForm = useCallback((): void => {
     console.log('preview form button clicked')
-  }, [])
+    navigate(`/${formId}/${PREVIEW_ROUTE}`)
```
Comment (karrui) [inline frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbarContainer.tsx]: nit: should we remove this log statement now that it's implemented? Even better, probably do not need to use `navigate`; we can directly use `react-router`'s `Link` element

File: frontend/src/features/admin-form/common/components/PreviewFormHeader/PreviewFormHeader.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+export type PreviewHeaderMode = 'preview'
+
+export const PreviewFormHeader = (): JSX.Element => {
+  const { formId } = useParams()
+
+  return (
+    <Flex bg="success.600" py="0.5rem" px="2rem" display="flex" width="100%">
+      <Flex align="center" flex={1} justify="space-between" flexDir="row">
+        <Box>
```
Comment (karrui) [inline frontend/src/features/admin-form/common/components/PreviewFormHeader/PreviewFormHeader.tsx]: nit: extraneous `Box`?

File: frontend/src/features/admin-form/preview/PreviewFormProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    /* enabled= */ !submissionData,
+  )
+  const { data: { captchaPublicKey } = {} } = useEnv(
+    /* enabled= */ !!data?.form.hasCaptcha,
+  )
+  const { hasLoaded, getCaptchaResponse, containerId } = useRecaptcha({
+    sitekey: data?.form.hasCaptcha ? captchaPublicKey : undefined,
+  })
+
+  const [cachedDto, setCachedDto] = useState<PreviewFormViewDto>()
```
Comment (karrui) [inline frontend/src/features/admin-form/preview/PreviewFormProvider.tsx]: I think all these (except `usePreviewForm`)  can also go into the `useCommonFormProvider` hook

File: frontend/src/features/admin-form/preview/PreviewFormProvider.tsx [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+  useTimeout(generateVfnExpiryToast, expiryInMs)
+
+  const showErrorToast = useCallback(() => {
+    toast({
+      status: 'danger',
+      description:
+        'An error occurred whilst processing your submission. Please refresh and try again.',
+    })
+  }, [toast])
```
Comment (karrui) [inline frontend/src/features/admin-form/preview/PreviewFormProvider.tsx]: All these too haha

File: frontend/src/features/admin-form/preview/PreviewFormProvider.tsx [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    } catch (error) {
+      if (error instanceof RecaptchaClosedError) {
+        // Do nothing if recaptcha is closed.
+        return
+      }
+      return showErrorToast()
+    }
+
+    setSubmissionData({ id: 'submissionId', timeInEpochMs: Date.now() })
+  }, [cachedDto, getCaptchaResponse, showErrorToast])
```
Comment (karrui) [inline frontend/src/features/admin-form/preview/PreviewFormProvider.tsx]: ah, for context, we actually have an endpoint for admins to preview and test their form submissions. It still goes to our servers and triggers the entire pipeline (with the exception of actually storing the response metadata). 

Can check it out [here](https://github.com/opengovsg/FormSG/blob/form-v2/develop/src/app/routes/api/v3/admin/forms/admin-forms.preview.routes.ts#L38) 

(so we still need mutations here, the only difference between preview and public form is the GET and POST endpoints. That's why we can more or less use all the methods except the queries, and maybe some UI differences)

File: frontend/src/features/admin-form/preview/PreviewFormProvider.tsx [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+    form.form_fields.forEach((f) => {
+      if (f.fieldType !== BasicField.Section) return
+      sections.push({
+        title: f.title,
+        _id: f._id,
+      })
+    })
+
+    return sections
+  }, [cachedDto, isAuthRequired])
```
Comment (karrui) [inline frontend/src/features/admin-form/preview/PreviewFormProvider.tsx:136]: can also go into the common hook

File: frontend/src/features/admin-form/preview/PreviewFormProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    formId,
+    // Stop querying once submissionData is present.
+    /* enabled= */ !submissionData,
+  )
+  const { data: { captchaPublicKey } = {} } = useEnv(
+    /* enabled= */ !!data?.form.hasCaptcha,
+  )
+  const { hasLoaded, getCaptchaResponse, containerId } = useRecaptcha({
+    sitekey: data?.form.hasCaptcha ? captchaPublicKey : undefined,
+  })
```
Comment (karrui) [inline frontend/src/features/admin-form/preview/PreviewFormProvider.tsx]: actually these may not need to be in the common hook, since we should disable captcha on preview hehe

File: frontend/src/features/public-form/PublicFormPage.tsx [hunk 12]
```diff
@@ -16,7 +16,7 @@ export const PublicFormPage = (): JSX.Element => {
       <FormStartPage />
       <PublicFormWrapper>
         <FormFields />
-        <FormEndPage />
+        <FormEndPage isPreview={false} />
```
Comment (karrui) [inline frontend/src/features/public-form/PublicFormPage.tsx]: nit: probably don't need this prop, since the absence of a prop is automatically falsey.

File: frontend/src/features/public-form/components/FormEndPage/FormEndPageContainer.tsx [hunk 13]
```diff
@@ -42,7 +56,9 @@ export const FormEndPageContainer = (): JSX.Element | null => {
         formTitle={form.title}
         endPage={form.endPage}
         isFeedbackSubmitted={isFeedbackSubmitted}
-        handleSubmitFeedback={handleSubmitFeedback}
+        handleSubmitFeedback={
+          isPreview ? handleSubmitFeedbackPreview : handleSubmitFeedback
+        }
```
Comment (karrui) [inline frontend/src/features/public-form/components/FormEndPage/FormEndPageContainer.tsx]: nit: should we just call the preview function in `handleSubmitFeedback` directly

File: frontend/src/features/admin-form/preview/PreviewFormProvider.tsx (hunk 9 above)
Comment (wanlingt) [reply frontend/src/features/admin-form/preview/PreviewFormProvider.tsx]: thank u! changed to use those endpoints :)

File: frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbarContainer.tsx (hunk 5 above)
Comment (wanlingt) [reply frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbarContainer.tsx]: oops good spot, have removed the log statement! I left it as `navigate` as I couldn't figure out how `Link` would work with the Preview Button's [onClick](https://github.com/opengovsg/FormSG/blob/form-v2/develop/frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbar.tsx#L157), advise pls :)

File: frontend/src/app/AppRouter.tsx [hunk 14]
```diff
@@ -90,6 +92,10 @@ export const AppRouter = (): JSX.Element => {
             />
           </Route>
         </Route>
+        <Route
+          path={`/:formId/${ADMINFORM_PREVIEW_ROUTE}`}
```
Comment (karrui) [inline frontend/src/app/AppRouter.tsx]:
```suggestion
          path={`${ADMINFORM_ROUTE}/:formId/${ADMINFORM_PREVIEW_ROUTE}`}
```

File: frontend/src/features/admin-form/common/AdminViewFormService.ts [hunk 15]
```diff
@@ -87,3 +97,44 @@ export const removeSelfFromFormCollaborators = async (
     `${ADMIN_FORM_ENDPOINT}/${formId}/collaborators/self`,
   )
 }
+
+/**
+ * Submit an email mode form in preview mode
+ */
+
```
Comment (karrui) [inline frontend/src/features/admin-form/common/AdminViewFormService.ts]: nit: empty line

File: frontend/src/features/admin-form/common/AdminViewFormService.ts [hunk 16]
```diff
@@ -87,3 +97,44 @@ export const removeSelfFromFormCollaborators = async (
...
+export const submitEmailModeForm = async ({
+  formFields,
+  formInputs,
+  formId,
+}: //   captchaResponse = null,
```
Comment (karrui) [inline frontend/src/features/admin-form/common/AdminViewFormService.ts]: should we remove this comment remnant?

File: frontend/src/features/admin-form/common/AdminViewFormService.ts [hunk 17]
```diff
@@ -87,3 +97,44 @@ export const removeSelfFromFormCollaborators = async (
     `${ADMIN_FORM_ENDPOINT}/${formId}/collaborators/self`,
   )
 }
+
+/**
+ * Submit an email mode form in preview mode
+ */
+
+export const submitEmailModeForm = async ({
```
Comment (karrui) [inline frontend/src/features/admin-form/common/AdminViewFormService.ts]: the name of this function seems a little too vague, should we rename it to `submitEmailModeFormPreview`?

File: frontend/src/features/admin-form/common/AdminViewFormService.ts [hunk 18]
```diff
@@ -87,3 +97,44 @@ export const removeSelfFromFormCollaborators = async (
...
+    `${ADMIN_FORM_ENDPOINT}/${formId}/preview/submissions/email`,
+    formData,
+  ).then(({ data }) => data)
+}
+
+/**
+ * Submit a storage mode form in preview mode
+ */
+
+export const submitStorageModeForm = async ({
```
Comment (karrui) [inline frontend/src/features/admin-form/common/AdminViewFormService.ts]: here too, also empty new line

File: frontend/src/features/admin-form/common/AdminViewFormService.ts [hunk 19]
```diff
@@ -87,3 +97,44 @@ export const removeSelfFromFormCollaborators = async (
...
+  formFields,
+  formInputs,
+  formId,
+  publicKey,
+}: // captchaResponse = null,
```
Comment (karrui) [inline frontend/src/features/admin-form/common/AdminViewFormService.ts]: also here

File: frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbarContainer.tsx [hunk 20]
```diff
@@ -52,8 +54,8 @@ const useAdminFormNavbar = () => {
   )
 
   const handlePreviewForm = useCallback((): void => {
-    console.log('preview form button clicked')
-  }, [])
+    navigate(`/${formId}/${ADMINFORM_PREVIEW_ROUTE}`)
+  }, [navigate, formId])
```
Comment (karrui) [inline frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbarContainer.tsx]: we can convert the button to a link instead! 

```tsx
// AdminFormNavbar.tsx

import {Link as ReactLink} from 'react-router-dom'

... 
<Button
  as={ReactLink}
  to={`${ADMINFORM_ROUTE}/${formId}/${ADMINFORM_PREVIEW_ROUTE}`}
  {...mobileDrawerExtraButtonProps}
  leftIcon={<BiShow fontSize="1.25rem" />}
>
    Preview form
</Button>
```

File: frontend/src/features/public-form/PublicFormContext.tsx [hunk 21]
```diff
@@ -36,8 +36,8 @@ export interface PublicFormContextProps
   /** Callback to be invoked when user submits public form. */
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   handleSubmitForm: (formInputs: any) => void
-  /** id of container to render captcha in  */
-  captchaContainerId: string
+  /** id of container to render captcha in. Optional in preview mode */
```
Comment (karrui) [inline frontend/src/features/public-form/PublicFormContext.tsx]: nit: probably don't need to tie this to preview mode so explicitly, can just say captcha will be instantiated if provided hehe

File: frontend/src/features/public-form/PublicFormProvider.tsx [hunk 22]
```diff
@@ -155,6 +144,82 @@ export const PublicFormProvider = ({
...
+    isNotFormId,
+    isFormNotFound,
+    cachedDto,
+    getTransactionId,
+    expiryInMs,
+    showErrorToast,
+    isAuthRequired,
+    sectionScrollData,
+    error,
+  } = useCommonFormProvider(formId, data, PublicFormError)
```
Comment (karrui) [inline frontend/src/features/public-form/PublicFormProvider.tsx]: supernit: could probably make code more readable by only doing a spread when passing in values to the context.

(unless they are also used somewhere else and that makes the code messier then we should not do it)

```tsx
const commonFormValues = useCommonFormProvider(formId, data, PublicFormError)

... 
<PublicFormContext.Provider
  value={{
    ... // other props,
    ...commonFormValues
  }}
>...
    
```

File: frontend/src/features/public-form/PublicFormProvider.tsx [hunk 23]
```diff
@@ -47,35 +50,23 @@ interface PublicFormProviderProps {
...
-export const PublicFormProvider = ({
-  formId,
-  children,
-}: PublicFormProviderProps): JSX.Element => {
-  // Once form has been submitted, submission data will be set here.
-  const [submissionData, setSubmissionData] = useState<SubmissionData>()
+export function useCommonFormProvider(
+  formId: string,
+  data: PreviewFormViewDto | PublicFormViewDto | undefined,
+  error: ApiError | null,
```
Comment (karrui) [inline frontend/src/features/public-form/PublicFormProvider.tsx]: any reason why this is required? if something relies on this then we should not pass this in haha.

Comment (karrui) [inline frontend/src/features/public-form/PublicFormProvider.tsx]: suggest not passing in any args to this hook, it should only contain the truly shared variables imho. Other implementations can probably be duplicated.

File: frontend/src/features/public-form/PublicFormProvider.tsx [hunk 24]
```diff
@@ -47,35 +50,23 @@ interface PublicFormProviderProps {
...
   const vfnToastIdRef = useRef<string | number>()
   const desyncToastIdRef = useRef<string | number>()
+  const isNotFormId = useMemo(() => !FORMID_REGEX.test(formId), [formId])
+
+  const [cachedDto, setCachedDto] = useState<
+    PreviewFormViewDto | PublicFormViewDto
+  >()
 
   useEffect(() => {
     if (data) {
```
Comment (karrui) [inline frontend/src/features/public-form/PublicFormProvider.tsx:128]: this should not be in the shared hook, otherwise there will be erroneous tracking of visits to public form even if it is a preview form. or we can add a param as to whether the current form is in preview mode...

File: frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbarContainer.tsx (hunk 20 above)
Comment (wanlingt) [reply frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbarContainer.tsx]: I've kept it as `IconButton` in `AdminFormNavbar` because I thought it would be easier to standardise the formatting that way (same as the "Manage Collaborators" `IconButton` beside it). Hence have left this as `navigate`

File: frontend/src/features/public-form/PublicFormProvider.tsx (hunk 22 above)
Comment (wanlingt) [reply frontend/src/features/public-form/PublicFormProvider.tsx]: added for the values which are only used in `PublicFormContext`

File: frontend/src/features/public-form/PublicFormProvider.tsx (hunk 23 above)
Comment (wanlingt) [reply frontend/src/features/public-form/PublicFormProvider.tsx]: makes sense, i've removed `data` and `error` but have left `formId` in

File: frontend/src/features/public-form/PublicFormProvider.tsx (hunk 24 above)
Comment (wanlingt) [reply frontend/src/features/public-form/PublicFormProvider.tsx:128]: removed from shared hook!
Separately, I've also removed the other functions that depend on `cachedDto` because I didn't want to keep passing `cachedDto` back and forth.

File: frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbar.tsx [hunk 25]
```diff
@@ -55,6 +57,7 @@ export const AdminFormNavbar = ({
 }: AdminFormNavbarProps): JSX.Element => {
   const { ref, onMouseDown } = useDraggable<HTMLDivElement>()
   const { isOpen, onClose, onOpen } = useDisclosure()
+  const { formId } = useParams()
```
Comment (karrui) [inline frontend/src/features/admin-form/common/components/AdminFormNavbar/AdminFormNavbar.tsx]: nit: unused variable?

File: frontend/src/features/admin-form/preview/PreviewFormPage.tsx [hunk 26]
```diff
@@ -0,0 +1,33 @@
+import { useParams } from 'react-router-dom'
+
+import GovtMasthead from '~components/GovtMasthead'
+
+import FormEndPage from '~features//public-form/components/FormEndPage'
+import FormFields from '~features//public-form/components/FormFields'
+import { FormFooter } from '~features//public-form/components/FormFooter'
+import FormStartPage from '~features//public-form/components/FormStartPage'
+import { PublicFormWrapper } from '~features//public-form/components/PublicFormWrapper'
```
Comment (karrui) [inline frontend/src/features/admin-form/preview/PreviewFormPage.tsx]: nit: extra slashes in route

File: frontend/src/features/admin-form/preview/PreviewFormPage.tsx [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+  if (!formId) throw new Error('No formId provided')
+
+  return (
+    <PreviewFormProvider formId={formId}>
+      <GovtMasthead />
+      <PreviewFormHeader />
+      <FormStartPage />
+      <PublicFormWrapper>
+        <FormFields />
+        <FormEndPage isPreview={true} />
```
Comment (karrui) [inline frontend/src/features/admin-form/preview/PreviewFormPage.tsx]: nit: presence of a prop is automatically truthy

```suggestion
        <FormEndPage isPreview />
```

File: frontend/src/features/public-form/PublicFormProvider.tsx [hunk 28]
```diff
@@ -47,20 +47,64 @@ interface PublicFormProviderProps {
...
+  const desyncToastIdRef = useRef<string | number>()
+
+  const getTransactionId = useCallback(async () => {
+    if (!vfnTransaction || isPast(vfnTransaction.expireAt)) {
+      const result = await createTransactionMutation.mutateAsync()
+      setVfnTransaction(result)
+      return result.transactionId
+    }
+    return vfnTransaction.transactionId
+  }, [createTransactionMutation, vfnTransaction])
```
Comment (karrui) [inline frontend/src/features/public-form/PublicFormProvider.tsx:66]: this might fail in preview mode, since the form could be private. we should have distinct functions between preview and public mode for verification

File: frontend/src/features/public-form/PublicFormProvider.tsx [hunk 29]
```diff
@@ -47,20 +47,64 @@ interface PublicFormProviderProps {
...
+
+  const isNotFormId = useMemo(() => !FORMID_REGEX.test(formId), [formId])
+
+  const expiryInMs = useMemo(() => {
+    if (!vfnTransaction?.expireAt) return null
+    return differenceInMilliseconds(vfnTransaction.expireAt, Date.now())
+  }, [vfnTransaction])
```
Comment (karrui) [inline frontend/src/features/public-form/PublicFormProvider.tsx:73]: also can probably be different for public/preview. might be able to be done in another PR

File: frontend/src/features/public-form/components/FormEndPage/FormEndPageContainer.tsx [hunk 30]
```diff
@@ -9,25 +9,45 @@ import { usePublicFormContext } from '~features/public-form/PublicFormContext'
...
+  /**
+   * Handles feedback submission
+   * @param isPreview whether form is in preview mode
+   */
   const handleSubmitFeedback = useCallback(
     (inputs: FeedbackFormInput) => {
+      // no mutation required in preview-form mode
+      if (isPreview) {
+        toast({
+          description: 'Thank you for submitting your feedback!',
```
Comment (karrui) [inline frontend/src/features/public-form/components/FormEndPage/FormEndPageContainer.tsx]: maybe we can update the toast to be more descriptive that this is a preview. 

```suggestion
          description: 'Thank you for submitting your feedback! Since you are in preview mode, the feedback is not stored.',
```

**Review by karrui — APPROVED** [review]

File: frontend/src/features/public-form/PublicFormProvider.tsx (hunk 28 above)
Comment (wanlingt) [reply frontend/src/features/public-form/PublicFormProvider.tsx:66]: Is the reason for distinct functions because this might fail at `createTransactionMutation`? Traced it to [this](https://github.com/opengovsg/FormSG/blob/form-v2/develop/src/app/modules/verification/verification.controller.ts#L32), from the APIs called it looks like there's no check on whether the form is public or private, just whether the formId exists in the DB.

Relatedly, how do we want the behaviour of the functions to be different for preview vs public verification?

Comment (wanlingt) [reply frontend/src/features/public-form/PublicFormProvider.tsx:66]: As discussed - to maintain feature parity with the current Angular version, will leave the current PR as it is (same function used for both public and private forms)

Will open a new issue for mocking transactions in preview mode (e.g. actual SMSes will not be sent)

File: frontend/src/features/public-form/PublicFormProvider.tsx (hunk 29 above)
Comment (wanlingt) [reply frontend/src/features/public-form/PublicFormProvider.tsx:73]: refer to above comment related to `getTransactionId `
